velodyne_gazebo_plugins/src/GazeboRosVelodyneLaser.cpp

```diff
@@ -89,151 +89,141 @@ GazeboRosVelodyneLaser::SampleAngleInterval(sdf::ElementPtr _sdf,
-  samples.reserve(_sample_count);
+  using ScanPattern = GazeboRosVelodyneLaser::ScanPattern;
+  const ignition::math::Angle kEpsilonAngle{1e-6};
   if (_sdf->HasElement("angles")) {
-    size_t index = 0;
+    std::vector<ScanPattern::Sample> samples;
+    samples.reserve(_sample_count);
     std::istringstream iss(_sdf->GetElement("angles")->Get<std::string>());
-    std::transform(std::istream_iterator<double>(iss), std::istream_iterator<double>(),
-                   std::back_inserter(samples), [&index] (double angle) {
-                     return GazeboRosVelodyneLaser::ScanPattern::Sample(index++, angle);
-                   });
-    if (static_cast<int>(samples.size()) == _sample_count) {
-      auto it = std::find_if(samples.begin(), samples.end(),
-                             [_min_angle, _max_angle] (const GazeboRosVelodyneLaser::ScanPattern::Sample & sample) {
-                               const ignition::math::Angle kEpsilonAngle{1e-3};
-                               return ((sample.angle < _min_angle - kEpsilonAngle) ||
-                                       (sample.angle > _max_angle + kEpsilonAngle));
-                             });
-      if (it == samples.end()) {
-        return samples;
-      } else {
+    for (auto it = std::istream_iterator<double>(iss); it != std::istream_iterator<double>(); ++it) {
+      ignition::math::Angle angle(*it);
+      if (angle > _max_angle + kEpsilonAngle || angle < _min_angle - kEpsilonAngle) {
         RCLCPP_WARN(ros_node_->get_logger(),
                     "Velodyne laser plugin given a %s scan angle out of bounds: "
                     "%.16g (%.16g deg) < %.16g rad (%.16g deg) < %.16g (%.16g deg), ignoring setting.",
-                    _min_angle.Radian(), _min_angle.Degree(), it->angle.Radian(), it->angle.Degree(),
-                    _max_angle.Radian(), _max_angle.Degree(), _sdf->GetName().c_str());
+                    _sdf->GetName().c_str(), _min_angle.Radian(), _min_angle.Degree(),
+                    angle.Radian(), angle.Degree(), _max_angle.Radian(), _max_angle.Degree());
+        return SampleAngleIntervalEvenly(_min_angle, _max_angle, _sample_count);
       }
-    } else {
-      RCLCPP_WARN(ros_node_->get_logger(),
-                  "Velodyne laser plugin not given %d %s scan angles, ignoring setting.",
-                  _sample_count, _sdf->GetName().c_str());
+      angle = angle >= _min_angle ? (angle <= _max_angle ? angle : _max_angle) : _min_angle;
+      double index = (_sample_count - 1) * (angle - _min_angle).Radian() / (_max_angle - _min_angle).Radian();
+      index = index >= 0 ? (index <= (_sample_count - 1) ? index : (_sample_count - 1)) : 0;
+      samples.emplace_back(index, angle);
     }
-  } else {
-    RCLCPP_WARN(ros_node_->get_logger(),
-                "Velodyne laser plugin not given any %s scan pattern, ignoring setting.",
-                _sdf->GetName().c_str());
+    samples.shrink_to_fit();
```

Comment:
Well, since we interpolate, the actual amount of samples that are of any interest may be smaller than the available amount i.e. finer grid, better interpolation. That's what we do for GPU rays actually -- note we double up the sample count in `VLP-32C.urdf.xacro`.